prov/efa/docs/efa_rdm_protocol_v4.md

```diff
@@ -0,0 +1,1318 @@
+Exactly how to achieve that is up to the implementation to decide, the easiest way to do that is to keep including
+raw address header in the EAGER_MSGRTM even after receiving the handshake packet.
+
+Note, because this is an extra request, an endpoint cannot assume its peer will comply with the request. Therefore,
+the receiving endpoint must be able to handle the situation that a received packet does not have the expected header length.
+
+In that case, implementation will have two choices:
+
+1. write a truncated error completion entry
+2. move the application data to right place.
+
+Note this extra request was initially introduced as an extra feature named "zero copy receive", but it is later realized
+that is not an feature because the peer does not do anything different. Rather, it is an expectation the receiving
+endpoint has for the sender. Therefore, it was re-interpreted as an extra request named "stale header length".
+This re-interpretation does not change the implementation, thus does not cause backward incompatibility.
+
+### 4.4 have connection ID in packet header (connid header)
+
+The "have connection ID in packet header" extra request was introduced with libfabric 1.14.0 release, and was
+assigned the ID 3.
+
+This extra feature is designed to solve the "QP collision" problem, which is commonly experienced in
+client-server type of application.
+
+The "QP collision" problem arise from fact that the EFA device uses the gid + qpn as the unique
+identifier of a peer. Recall that raw address of EFA endpoint is consisted of 3 parts:
+gid + qpn + connid. EFA device only recognizes gid and qpn. The connid was generated by the endpoint
+itself during its initialization.
+
+Because of that, it is possible for an endpoint to receive packets from a destroyed
+peer. The endpoint does not know how to handle such a packet, and will crash or cause data corruption.
+
+The problem can be demonstrated using the following workflow:
+
+    1. We have two instances A and B.
+    2. On instance A we create endpoint 1.
+    3. On instance B we create endpoint 2.
+    4. Endpoint 1 and 2 start communication with each other.
+    5. There is an accident on instance B, endpoint 2 was destroyed.
+    6. Before it is destroyed, endpoint 2 sent some packets to 1.
+    7. On instance B, a new endpoint 3 was created, which has the same QPN as endpoint 2.
+    8. Endpoint 3 start sending packets to endpoint 1.
+    9. Upon receive packets from endpoint 3, endpoint 1 recognize this is a new peer via its connection ID.
+       and will insert endpoint 3's address to address vector.
+    10. The packets from endpoint 2 (in step 6) arrived, because endpoint 2 has the same gid+qpn as endpoint 3.
+        EFA device will think these packets are from endpoint 3.
+    11. Endpoint 1 will not be able to process these packets, and will crash.
+
+The solution to this problem is to include sender's connid in each packet's header. In this case,
```

Comment:
Does it make sense to put connid in the base header instead so we don't have to add it to each packet type (now and in the future)? Or is it simpler to not do that?

Reply:
I think it is simpler to add it to each packet type, and using a bit in `flags` to indicate its existence in the header.

To add it to base header would require introduce new packet types. Because current packet header are all based on 4 bytes base header, and the alignment would change.

Reply:
Got it, makes sense.